### PR TITLE
Implementation

### PR DESCRIPTION
**Pull request by fazo96**
@kulpreet you can turn on Travis so that we can get it to run the tests

Checklist:

- [x] basic API implementation
- [x] some minor API calls are still missing to match ChluIPFS
- [x] is able to manage the DID
- [x] persistence of the DID
- [x] unit tests
- [x] integration testing (requires backend)
- [x] signs DID when publishing
- [x] signs review records when publishing
- [ ] run tests in browser
- [ ] browser build with webpack

The implementation has been done by porting code over from ChluIPFS (or importing pieces from it). It's not ideal, but it was fast and we can iterate over it to improve

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        return await this.api.readReviewRecord(multihash, options)
+    }
+
+    async storeReviewRecord(reviewRecord, options = {}) {
+        // prepare the review record
+        // this will sign it as both customer and issuer. The server can then replace the issuer signature with its own
+        // the 'validate' option is forced to false because the API Client cannot validate.
+        const bitcoinTransactionHash = get(options, 'bitcoinTransactionHash', null)
+
+        reviewRecord.issuer = await this.api.getPublisherDID() // Authorize server to issue review
```
Comment (kulpreet) [inline src/index.js:55]: Yeah, we should take away any reference to PoPR backed reviews in this store method.

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    async getReviewsAboutDID(didId, offset, limit) {
+        return await this.api.getReviewsAboutDID(didId, offset, limit)
+    }
+
+    async getDID(didId, waitUntilPresent = false) {
+        return await this.didIpfsHelper.getDID(didId, waitUntilPresent)
+    }
+
+    async getPoPR(multihash) {
```
Comment (kulpreet) [inline src/index.js:75]: Take this out?

File: src/modules/didIpfsHelper.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
+const ChluDID = require('chlu-did/src')
```
Comment (kulpreet) [inline src/modules/didIpfsHelper.js:1]: This is getting duplicated from the chlu-ipfs, right? Just clarifying.

File: src/modules/http.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    async storeReviewRecord(reviewRecord, opt = {}) {
+        const options = Object.assign({}, {
+            publish: true
+        }, opt)
+        const response = await this.axios.post(`${this.publishApiUrl}/reviews`, reviewRecord, {
+            params: {
+                publish: get(options, 'publish', true),
+                bitcoinTransactionHash: get(options, 'bitcoinTransactionHash', ''),
```
Comment (kulpreet) [inline src/modules/http.js:35]: Remove transactionhash referernces

File: src/modules/http.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    async getDID(didId, waitUntilPresent = false) {
+        const response = await this.axios.get(`${this.queryApiUrl}/dids/${didId}`, {
+            params: {
+                waitUntilPresent: Boolean(waitUntilPresent)
+            }
+        })
+        return response.data
+    }
+
+    async getPoPR(multihash) {
```
Comment (kulpreet) [inline src/modules/http.js:66]: Remove?

File: src/utils/ipfs.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+const DAGNode = require('ipld-dag-pb').DAGNode;
```
Comment (kulpreet) [inline src/utils/ipfs.js:1]: I wonder if to store/get/resolve reviews can go via the api/client/server?

I am thinking only for the crawled reviews, or to say, during airdrop case.

That way, the wallets during airdrops, don't have to start an ipfs instance? Does it make sense?

**Review by kulpreet — CHANGES_REQUESTED** [review]
Refactoring questions

File: src/modules/didIpfsHelper.js (hunk 3 above)
Comment (fazo96) [reply src/modules/didIpfsHelper.js:1]: It's not duplicated, I made changes here. But most of the code is a duplicate, so when we refactor the DID stuff we will fix this and make sure we don't have repeated code.

File: src/modules/http.js (hunk 5 above)
Comment (fazo96) [reply src/modules/http.js:66]: We need this in the wallet. Or we can have the wallet get it from IPFS, what do you think is best?

File: src/modules/persistence.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
+
```
Comment (fazo96) [inline src/modules/persistence.js:1]: Just noting that this module is also a duplicate from chluIpfs, with changes made here.

File: src/utils/ipfs.js (hunk 6 above)
Comment (fazo96) [reply src/utils/ipfs.js:1]: They have to start it for publishing. But for reading, they can use the API. We can figure out which to use later, they both work.

Comment (fazo96) [reply src/utils/ipfs.js:1]: Remember that this 'ipld-dag-pb' module does not require IPFS to be running at all! It's just to serialize the data and calculate the multihash

File: src/index.js (hunk 2 above)
Comment (fazo96) [reply src/index.js:75]: See comment below

**Pull request merged**